File: Grafowe_modelowanie_systemow/najkrotsze_sciezki_EULER.cpp
#include <iostream>
#include <string>
#include <limits>

using namespace std;

int main()
{
	int numberOfTests; // how many times to run code (#of Test Cases)
	int numberOfCities; // how many cities per Test Case 

	string* cityNames; // array for: city -> name
	int*** cityNeighbors; // 3d array for: city -> neighbor array -> neighbor data
	int** distances; // 2d array for: city -> distance + visited (2 cell array)
	int* howManyNeighbors; // array for: city -> how many neighbors (for for loops)

	string startCity;
	string endCity;
	string cityName;
	int numberOfNeighbors;
	int numberOfPaths;

	cin >> numberOfTests;
	for (int ii = 0; ii < numberOfTests; ii++){
		cin >> numberOfCities;
		cityNames = new string[numberOfCities];
		cityNeighbors = new int**[numberOfCities];
		distances = new int*[numberOfCities];
		howManyNeighbors = new int[numberOfCities];
		// LOAD CITIES DATA
		for (int jj = 0; jj < numberOfCities; jj++){
			distances[jj] = new int[2]; // [0]:distance [1]: isVisited -> 0: no, 1: yes
			cin >> cityName;
			cityNames[jj] = cityName;
			cin >> numberOfNeighbors;
			cityNeighbors[jj] = new int*[numberOfNeighbors];
			howManyNeighbors[jj] = numberOfNeighbors;
			for (int kk = 0; kk < numberOfNeighbors; kk++){
				cityNeighbors[jj][kk] = new int[2];
				cin >> cityNeighbors[jj][kk][0]; 
				cityNeighbors[jj][kk][0]--; // acount for cities numbered from 1 not 0
				cin >> cityNeighbors[jj][kk][1]; // distance to neighbor city
			}
		}
		// LOAD PATH DATA AND RUN DIJKSTRA ON THEM
		cin >> numberOfPaths;
		for (int nn = 0; nn < numberOfPaths; nn++){
			cin >> startCity;
			cin >> endCity;
			int startIndex = 0;
			int endIndex = 0;
			int lowestDistance;
			int lowestDistanceIndex;
			for (int ll = 0; ll < numberOfCities; ll++){
				if (cityNames[ll] == startCity) {
					startIndex = ll;
				}
				if (cityNames[ll] == endCity) {
					endIndex = ll;
				}
				// reset all distances and visited
				distances[ll][0] = numeric_limits<int>::max(); // set distance as 'infinity'
				distances[ll][1] = 0; // nothing visited yet
			}
			distances[startIndex][0] = 0; // set source
			for (int tt = 0; tt < numberOfCities; tt++){ // go through all verticies
				lowestDistance = numeric_limits<int>::max();
				for (int uu = 0; uu < numberOfCities; uu++){ // find lowest value, not visited vertix
					if (distances[uu][1] == 0 && distances[uu][0] < lowestDistance){
						lowestDistance = distances[uu][0];
						lowestDistanceIndex = uu;
					}
				}
				for (int mm = 0; mm < howManyNeighbors[lowestDistanceIndex]; mm++){ // for each neighbor
					int neighborIndex = cityNeighbors[lowestDistanceIndex][mm][0];
					int neighborDistance = cityNeighbors[lowestDistanceIndex][mm][1];
					if (distances[neighborIndex][0] > (distances[lowestDistanceIndex][0] + neighborDistance)){
						distances[neighborIndex][0] = (distances[lowestDistanceIndex][0] + neighborDistance); // update to shorter
					}
				}
				distances[lowestDistanceIndex][1] = 1; // mark city as visited
			}
			cout << distances[endIndex][0] << endl;
		}
		// clean up the memory
		delete[] cityNames;
		for (int i = 0; i < numberOfCities; i++){
			delete[] distances[i];
			for (int j = 0; j < howManyNeighbors[i]; j++){
				delete cityNeighbors[i][j];
			}
		}
		delete[] cityNeighbors;
		delete[] distances;
		delete[] howManyNeighbors;
	}
	return 0;
}

File: Grafowe_modelowanie_systemow/najkrotsze_sciezki_EULER.py
class city():
    def __init__(self):
        self.name = ""
        self.index = -1
        self.paths = {} # KEY target city (indexed by city index starting at 1) VALUE: int, cost to get to target (edge value)
        self.distanceFromSource = 0
        self.isVisited = False

    def getIndex(self):
        return self.index

    def setIndex(self, index):
        self.index = index

    def setVisited(self, boolean):
        self.isVisited = boolean

    def getVisited(self):
        return self.isVisited

    def setDistanceFromSource(self, distance):
        self.distanceFromSource = distance

    def getDistanceFromSource(self):
        return self.distanceFromSource

    def setCityAsSource(self):
        self.distanceFromSource = 0
        self.isVisited = False

    def resetCitySettings(self):
        self.distanceFromSource = float('inf')
        self.isVisited = False

    def addPath(self, destination, cost):
        self.paths[int(destination)] = int(cost)

    def getCost(self, destination):
        return self.paths[int(destination)]

    def getDestinations(self):
        return self.paths.keys()

    def setName(self, name): # might not need name at all if name stored in dictionary name -> index
        self.name = name

    def getName(self):
        return self.name

    def __str__(self):
        return "{} -> keys: {}, values: {}, distance: {}".format(self.name, self.paths.keys(), self.paths.values(), self.distanceFromSource)

class cityList():
    def __init__(self):
        self.cities = {} # dic of city objects -> key: city.index, value city object
        self.cityNames = {} # when adding a city to self.cities also set city.index as value HERE and city.name as KEY here
        self.start = None
        self.end = None

    def dijkstra(self, startName, endName):
        """
1. Start @ source and set distance from source to 0, other to infinity.
2. Pick lowest distance value from vertices not yet visited
3. update distances to their neighbours (if path from s->u->v is shorter than s->v)
4. repeat untill ALL verticies have been visited
        """
        self.start = self.cities[self.cityNames[startName]]
        self.end = self.cities[self.cityNames[endName]]
        for val in self.cities.values():
            val.resetCitySettings()
        self.start.setCityAsSource()
        while True:
            citiesToVisit = list(filter(lambda x: x.getVisited() == False, self.cities.values())) # get list of not visitied verticies
            #print("len(citiesToVisit) = {}".format(len(citiesToVisit)))
            if len(citiesToVisit) == 0:
                break
            lowestCostCity = min(citiesToVisit, key=lambda x: x.getDistanceFromSource())
            for neighbour in lowestCostCity.getDestinations():
                neighbourCost = self.cities[neighbour].getDistanceFromSource()   # distance neighbour already has
                alternativePath = lowestCostCity.getDistanceFromSource() + lowestCostCity.getCost(neighbour) # path from source to neighbour via current
                # neighbours distance from source VS current point + path to neighbour from current point
                if neighbourCost > alternativePath:
                    self.cities[neighbour].setDistanceFromSource(alternativePath)
                #self.debugPrintDistances()
            lowestCostCity.setVisited(True)
        return self.end.getDistanceFromSource()
        
    def loadCities(self):
        numberOfCities = int(input())
        for i in range(numberOfCities):
            newCity = city()
            cityName = input()
            cityIndex = i + 1
            newCity.setName(cityName)
            newCity.setIndex(cityIndex)
            numberOfNeighbours = int(input())
            for j in range(numberOfNeighbours):
                neighbourElements = list(input().split(' '))
                newCity.addPath(int(neighbourElements[0]), int(neighbourElements[1]))
            self.cityNames[cityName] = newCity.getIndex()
            self.cities[newCity.getIndex()] = newCity

    def findPaths(self):
        numberOfPaths = int(input())
        for i in range(numberOfPaths):
            targetList = input().split(' ')
            print(self.dijkstra(targetList[0], targetList[1]))

    def debugPrintDistances(self):
        for ele in self.cities.values():
            print(ele, end = '; ')
            print(ele.getDistanceFromSource())
                    


def runInputProgram():
    numberOfTests = int(input())
    for i in range(numberOfTests):
        currentCityList = cityList()
        currentCityList.loadCities()
        currentCityList.findPaths()
        if i < numberOfTests - 1:
            input() # read the delimiting newline
        
runInputProgram()
